File: packages/cli/commands/import.ts
import { ingestBundle, ingestGithub } from "@compositz/core";
import { dim, green, red } from "@std/fmt/colors";
import { storeDir } from "../lib.ts";

const USAGE =
  "usage: compositz import <archive.tar|archive.tar.gz|dir|github:owner/repo[/subdir][@ref]>";

/** Import a recipe → create an instance: a tar/tar.gz archive, a directory, or GitHub. */
export async function importCmd(args: string[]): Promise<number> {
  const arg = args[0];
  if (!arg) {
    console.error(red(USAGE));
    return 1;
  }

  let instance;
  if (arg.startsWith("github:")) {
    // A GitHub source: parse → download the codeload tarball → ingest (errors from
    // here propagate to main.ts and are printed in red).
    instance = await ingestGithub(arg, storeDir());
  } else {
    let stat: Deno.FileInfo;
    try {
      stat = await Deno.stat(arg);
    } catch {
      console.error(red(`not found: ${arg}`));
      return 1;
    }
    if (stat.isDirectory) {
      instance = await ingestBundle({ kind: "dir", dir: arg }, storeDir(), {
        source: `dir:${arg}`,
      });
    } else {
      // Stream the file through extraction (never buffer it whole in RAM).
      const file = await Deno.open(arg, { read: true });
      instance = await ingestBundle(
        { kind: "archive", stream: file.readable },
        storeDir(),
        { source: `file:${arg}` },
      );
    }
  }

  console.log(green(`OK — imported ${instance.manifest.name}`) + dim(` as ${instance.instanceId}`));
  console.log(dim(`  run: compositz up ${instance.instanceId}`));
  return 0;
}

File: crates/cli/src/commands/import.rs
//! `import` — ingest a recipe (tar/tar.gz archive, directory, or GitHub) into a
//! new instance, then deconflict its host ports against the other instances.

use anyhow::Result;
use compositz_core::{
    BundleSource, GithubIngestOpts, IngestOpts, Instance, deconflict_host_ports, ingest_bundle,
    ingest_github,
};

use crate::cli::store_dir;
use crate::style::{dim, green, red, yellow};

/// Import from a `github:owner/repo[/subdir][@ref]` source, a directory, or a
/// tar/tar.gz archive file. A missing local path prints a plain `not found: …`
/// (parity with the Deno import — no `error:` prefix) and exits 1.
pub async fn run(source: String) -> Result<i32> {
    let store = store_dir()?;

    let instance = if source.starts_with("github:") {
        ingest_github_blocking(source, store.clone()).await?
    } else {
        // Stat here (not in the blocking task) so a missing path yields the friendly
        // message + exit 1 rather than bubbling to the top-level `error:` handler.
        let Ok(metadata) = std::fs::metadata(&source) else {
            eprintln!("{}", red(&format!("not found: {source}")));
            return Ok(1);
        };
        ingest_path_blocking(source, store.clone(), metadata.is_dir()).await?
    };

    // Reassign any host port that collides with another instance's DEFINED port,
    // and say so (parity with the Deno import's deconflict-and-notify).
    for bump in deconflict_host_ports(&store, &instance)? {
        println!(
            "{}",
            yellow(&format!(
                "  note: {} port {} in use → reassigned to {}",
                bump.name, bump.from, bump.to
            ))
        );
    }

    println!(
        "{}{}",
        green(&format!("OK — imported {}", instance.manifest.name)),
        dim(&format!(" as {}", instance.instance_id))
    );
    println!(
        "{}",
        dim(&format!("  run: compositz up {}", instance.instance_id))
    );
    Ok(0)
}

/// Download + ingest a GitHub source off the async runtime (the whole fetch →
/// gunzip → untar pipeline is blocking).
async fn ingest_github_blocking(source: String, store: String) -> Result<Instance> {
    tokio::task::spawn_blocking(move || -> Result<Instance> {
        Ok(ingest_github(&source, &store, GithubIngestOpts::default())?)
    })
    .await?
}

/// Ingest a directory or an archive file off the async runtime (extraction / copy
/// is blocking). `is_dir` was determined by the caller's stat.
async fn ingest_path_blocking(source: String, store: String, is_dir: bool) -> Result<Instance> {
    tokio::task::spawn_blocking(move || -> Result<Instance> {
        let instance = if is_dir {
            ingest_bundle(
                BundleSource::Dir {
                    dir: source.clone(),
                },
                &store,
                IngestOpts {
                    source: Some(format!("dir:{source}")),
                    ..Default::default()
                },
            )?
        } else {
            // Stream the file through extraction (never buffer it whole in RAM).
            let file = std::fs::File::open(&source)?;
            ingest_bundle(
                BundleSource::Archive {
                    reader: Box::new(file),
                    subdir: None,
                },
                &store,
                IngestOpts {
                    source: Some(format!("file:{source}")),
                    ..Default::default()
                },
            )?
        };
        Ok(instance)
    })
    .await?
}
